fix(core): renderer-to-renderer2 migration not migrating methods

devversion opened the pull request:
The `renderer-to-renderer2` migration currently does not work
properly in v9 because the migration relies on the type checker
for detecting references to `Renderer` from `@angular/core`.

This is contradictory since the `Renderer` is no longer
exported in v9 `@angular/core`. In order to make sure that
the migration still works in v9, and that we can rely on the
type checker for the best possible detection, we take advantage
of module augmentation and in-memory add the `Renderer` export
to the `@angular/core` module.

---

Before coming up with the augmentation solution, I spent quite some time re-working the
detection to be not based on the `ts.TypeChecker`. Though, I went with this solution as it gives us
the best detection of the `Renderer` and doesn't involve reworking the whole detection.

crisbeto commented on packages/core/schematics/migrations/renderer-to-renderer2/index.ts:
```diff
@@ -44,8 +43,23 @@ export default function(): Rule {
+      return `
+        import '@angular/core';
+        declare module "@angular/core" {
+          class Renderer {}
```
We aren't compiling the code so maybe it doesn't matter, but wouldn't this technically cause compilation errors for something like `someInstance.setElementClass` since the `Renderer` class is empty? Could we do `declare const Renderer: any` instead?

devversion replied:
Yeah it would cause errors, but since we don't collect diagnostics, the error won't be reported. It shouldn't have an effect on the migration either.

 I think switching to `any` will break since the type checker will then return `any` if we call `getTypeAtLocation`.

crisbeto reviewed (approved):
LGTM

kara reviewed (approved):
LGTM

kara commented:
[presubmit](http://test/OCL:278669830:BASE:278669896:1572981669913:9613126f)

devversion committed:
fix(core): renderer-to-renderer2 migration not migrating methods

The `renderer-to-renderer2` migration currently does not work
properly in v9 because the migration relies on the type checker
for detecting references to `Renderer` from `@angular/core`.

This is contradictory since the `Renderer` is no longer
exported in v9 `@angular/core`. In order to make sure that
the migration still works in v9, and that we can rely on the
type checker for the best possible detection, we take advantage
of module augmentation and in-memory add the `Renderer` export
to the `@angular/core` module.

devversion commented on packages/core/schematics/utils/typescript/compiler_host.ts:
```diff
@@ -10,15 +10,18 @@ import {relative} from 'path';
-    const buffer = tree.read(relative(basePath, fileName));
+    const treeRelativePath = relative(basePath, fileName);
+    const fakeOutput = fakeRead ? fakeRead(treeRelativePath) : null;
+    const buffer = fakeOutput !== null ? fakeOutput : tree.read(treeRelativePath);
```
Updated.

devversion commented:
Caretaker: Failure for `integration_test` seems unrelated to this PR. I rebased on top of 44623a116158c8048ce61a873ed63a290c039202 (latest commit in master).

atscott commented:
Presubmit failures are also unrelated. Still need to investigate the integration_test failures
